compiler/rustc_trait_selection/src/transmute.rs

```diff
@@ -0,0 +1,78 @@
+use rustc_infer::infer::InferCtxt;
+use rustc_infer::traits::{Obligation, ObligationCause};
+use rustc_middle::ty::{Binder, ParamEnv, PolyTraitRef, Ty, TyCtxt};
+use rustc_transmute::layout::rustc::Ref;
+use rustc_transmute::{Answer, Assume, Condition, Reason};
+
+/// Attempts to confirm the given transmutability candidate.
+///
+/// The return type of this routine reflects that it is designed to support the
+/// needs of both trait selection and error reporting; it returns:
+/// - `Ok(None)` if the types are transmutable
+/// - `Ok(Some(src, dst, assumptions, condition))` if the `src` is transmutable
+///   into `dst` when `condition` holds under `assumptions`
+/// - `Err(None)` if transmutability cannot be assessed; e.g., due to a
+///   malformed `Assume`
+/// - `Err(Some((src, dst, reason))` if `src` is not transmutable into `dst`
+///   because of `reason`.
+pub(crate) fn confirm_transmutability_candidate<'tcx, P>(
+    infcx: &InferCtxt<'tcx>,
+    obligation: &Obligation<'tcx, P>,
+    trait_ref: PolyTraitRef<'tcx>,
+) -> Result<
+    Option<(Ty<'tcx>, Ty<'tcx>, Assume, Condition<Ref<'tcx>>)>,
+    Option<(Ty<'tcx>, Ty<'tcx>, Reason<Ref<'tcx>>)>,
+> {
+    /// Attempts to deeply normalize `ty`.
+    fn try_normalize<'tcx>(
```

Comment:
Can you give me an example of where this normalization is necessary? We should expect the obligation's predicate to already be normalized by the time we get confirmation.

Reply:
Specifically, I don't see any changes to tests that would necessitate this additional normalization.

Reply:
Actually, see top-level comment.